### ISSUE TITLE
JavaCompile fails for custom VM arguments passed via jvmArgs in external-javac (forked mode with javaHome != null)

### ISSUE THREAD

**Issue by dweiss** [gradle/gradle#22746]
TL;DR; The handling of VM arguments differs between argument provides in JavaCompile implementations.

The "standard" javac compiler accepts VM arguments via ```-J``` options. So, to pass ```-XX:+PrintFlagsFinal```, you'd add ```-J-XX:+PrintFlagsFinal``` to javac. Gradle explicitly prevents those "-J" options to appear in compiler arguments:

https://github.com/gradle/gradle/blob/master/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/JavaCompilerArgumentsBuilder.java#L115-L118

instead, it is suggested to pass them via ```forkOptions.jvmArgs```. This works in the in-process compiler mode and in the forked compiler mode but NOT with an external javac. Here is an example:

```
import org.gradle.internal.jvm.Jvm

plugins {
  id "java-library"
}

tasks.withType(JavaCompile) { JavaCompile task ->
  // pass a custom VM option to java compiler.
  task.options.forkOptions.jvmArgs += [
    '-XX:+PrintFlagsFinal'
  ]

  switch (project.hasProperty("mode") ? project.getProperty("mode") : "inprocess") {
    case "inprocess":
      // nothing.
      break;

    case "fork":
      task.options.fork = true
      break;

    case "fork+javaHome":
      task.options.fork = true
      task.options.forkOptions.javaHome=Jvm.current().javaHome
      break;
  }
}
```

When you run the above passing "-Pfork+javaHome", it will break because VM options are passed without the '-J' option:
![image](https://user-images.githubusercontent.com/199470/202127661-29b1c7ef-dd0a-4fb5-b0de-0a52acb6c5de.png)

On the other hand, adding "-J" to the arguments passed in jvmArgs breaks the "fork" mode (in-process ignores the attribute, that's fine). 

In other words, there is no single way to configure all three compilation modes consistently.

**Comment by dweiss** [comment]
[gradle-javac-bug.zip](https://github.com/gradle/gradle/files/10019590/gradle-javac-bug.zip)

A simple repro project (code above).

**Comment by eskatos** [comment]
Thank you for providing a valid reproducer.

The issue is in the backlog of the relevant team but the existence of a workaround makes it non-critical so it might take a while before a fix is made.

----

Gradle probably should append `-J` in the `fork+javaHome` case.
Declaring your arguments properly depending on the "fork mode" is the recommended workaround.

Note that future Gradle versions will reduce the discrepancy anyway by detecting proper Java toolchains in more cases, reducing the need for fully forking compilation to a `javac` executable.

**Comment by dweiss** [comment]
> reducing the need for fully forking compilation to a javac executable.

This applies to 99.9% of cases but not to people who live on the bleeding edge of JDK development, sadly. There's a delay before new JDK versions are detected/ recognized/ accepted and this delay is the reason why we (Apache Lucene) resort to custom compilation/javadoc invocations.
